server/src-lib/Hasura/GraphQL/Explain.hs

```diff
@@ -128,30 +126,23 @@ explainGQLQuery
   -> GQLExplain
   -> m EncJSON
 explainGQLQuery pool iso sc sqlGenCtx (GQLExplain query userVarsRaw)= do
-  (gCtx, _) <- flip runStateT sc $ getGCtx (userRole userInfo) gCtxMap
-  queryParts <- runReaderT (GV.getQueryParts query) gCtx
-  let topLevelNodes = TH.getTopLevelNodes (GV.qpOpDef queryParts)
-
-  unless (allHasuraNodes gCtx topLevelNodes) $
-    throw400 InvalidParams "only hasura queries can be explained"
-
-  (opTy, selSet) <- runReaderT (GV.validateGQ queryParts) gCtx
-  unless (opTy == G.OperationTypeQuery) $
-    throw400 InvalidParams "only queries can be explained"
-  let tx = mapM (explainField userInfo gCtx sqlGenCtx) (toList selSet)
-  plans <- liftIO (runExceptT $ runTx tx) >>= liftEither
-  return $ encJFromJValue plans
-
+  execPlan <- E.getExecPlan userInfo sc query
+  (gCtx, rootSelSet) <- case execPlan of
+    E.GExPHasura gCtx rootSelSet ->
+      return (gCtx, rootSelSet)
+    E.GExPRemote _ _  ->
+      throw400 InvalidParams "only hasura queries can be explained"
+  case rootSelSet of
+    GV.RQuery selSet -> do
+      let tx = mapM (explainField userInfo gCtx sqlGenCtx) (toList selSet)
+      plans <- liftIO (runExceptT $ runTx tx) >>= liftEither
+      return $ encJFromJValue plans
+    GV.RMutation _ ->
```

Comment:
we can use wildcard pattern match `_` if its not `RQuery`?

Reply:
It helps to list all the available constructors sometimes. For example, if we happen to introduce a new constructor later, with `_` we get no help from the compiler.